backend/msgpack/conn.go

```diff
@@ -146,8 +150,10 @@ func (c *connection) connectWithLock() error {
 
 func (c *connection) checkReconnectWithLock() bool {
 	// If we haven't accumulated enough failures to warrant another reconnect
-	// empty, simply returning false without reconnecting.
-	if c.numFailures <= c.threshold {
+	// and we haven't past the maximum duration since the last time we attempted
+	// to connect then we simply return false without reconnecting.
+	lastConnect := time.Unix(0, c.lastConnectNanos)
+	if c.numFailures <= c.threshold && c.nowFn().Before(lastConnect.Add(c.maxDuration)) {
```

Comment:
Might be more efficient to do below, can reuse `nowNanos` in `connectWithLock` too.
```
func (c *connection) nowNanos() int64 {
  return c.nowFn().UnixNano()
}

func (c *connection) checkReconnectWithLock() bool {
  ...
  if c.numFailures <= c.threshold && c.nowNanos() < c.lastConnectNanos + int64(c.maxDuration) {
    return false
  }
  ...
}
```